skiko/src/jvmTest/kotlin/org/jetbrains/skiko/SeveralClassloaders.kt

```diff
@@ -0,0 +1,84 @@
+package org.jetbrains.skiko
+
+import org.junit.Test
+import java.net.URL
+import java.net.URLClassLoader
+import kotlin.concurrent.thread
+
+private class PlatformAndURLClassLoader(classpath: List<URL>) :
+    ClassLoader(getPlatformClassLoader()) {
+    private val childClassLoader: ChildURLClassLoader
+
+    private class FindClassClassLoader(parent: ClassLoader?) : ClassLoader(parent) {
+        @Throws(ClassNotFoundException::class)
+        public override fun findClass(name: String): Class<*> {
+            return super.findClass(name)
+        }
+    }
+
+    private class ChildURLClassLoader(urls: Array<URL>?, private val realParent: FindClassClassLoader) :
+        URLClassLoader(urls, null) {
+        @Throws(ClassNotFoundException::class)
+        public override fun findClass(name: String): Class<*> {
+            return try {
+                super.findClass(name)
+            } catch (e: ClassNotFoundException) {
+                realParent.loadClass(name)
+            }
+        }
+    }
+
+    @Synchronized
+    @Throws(ClassNotFoundException::class)
+    override fun loadClass(name: String, resolve: Boolean): Class<*> {
+        return try {
+            childClassLoader.findClass(name)
+        } catch (e: ClassNotFoundException) {
+            super.loadClass(name, resolve)
+        }
+    }
+
+    init {
+        val urls = classpath.toTypedArray()
+        childClassLoader = ChildURLClassLoader(
+            urls, FindClassClassLoader(
+                parent
+            )
+        )
+    }
+}
+
+private fun testSkikoLoad(loader: ClassLoader) {
+    val clazz = loader.loadClass("org.jetbrains.skiko.LibraryTestImpl")
+    val tester = clazz.getDeclaredConstructor().newInstance()
+    val ptr = clazz.getMethod("run").invoke(tester) as Long
+    assert(ptr != 0L)
+}
+
+class SeveralClassloaders {
+    @Test
+    fun `load skiko in several classloaders`()  {
```

Comment:
Fails on OpenJDK 15, Window 10:

```
java.lang.ClassNotFoundException: org.jetbrains.skiko.LibraryTestImpl
	at java.base/java.lang.ClassLoader.findClass(ClassLoader.java:719)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at org.jetbrains.skiko.PlatformAndURLClassLoader.loadClass(SeveralClassloaders.kt:37)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at org.jetbrains.skiko.SeveralClassloadersKt.testSkikoLoad(SeveralClassloaders.kt:52)
	at org.jetbrains.skiko.SeveralClassloadersKt.access$testSkikoLoad(SeveralClassloaders.kt:1)
	at org.jetbrains.skiko.SeveralClassloaders.load skiko in several classloaders(SeveralClassloaders.kt:80)
```

Reply:
CI fails as well, hopefully fixed.

Reply:
Still fails :(

Methos Library.load doesn't call.

Reply:
Replacing `file://$jar` by `file:/$jar` helps, but there is another exception:
```
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.jetbrains.skiko.SeveralClassloadersKt.testSkikoLoad(SeveralClassloaders.kt:54)
	at org.jetbrains.skiko.SeveralClassloadersKt.access$testSkikoLoad(SeveralClassloaders.kt:1)
	at org.jetbrains.skiko.SeveralClassloaders.load skiko in several classloaders(SeveralClassloaders.kt:83)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.UnsatisfiedLinkError: C:\Users\Igor\AppData\Local\Temp\skiko14933663840593974363: Can't find dependent libraries
	at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
	at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:383)
	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:227)
	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:169)
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2407)
	at java.base/java.lang.Runtime.load0(Runtime.java:747)
	at java.base/java.lang.System.load(System.java:1857)
	at org.jetbrains.skiko.Library.loadLibraryOrCopy(Library.kt:26)
	at org.jetbrains.skiko.Library.load(Library.kt:60)
	at org.jetbrains.skija.impl.Library$Companion.staticLoad(Library.kt:12)
	at org.jetbrains.skija.impl.Library.staticLoad(Library.kt)
	at org.jetbrains.skija.Bitmap.<clinit>(Bitmap.java:10)
	at org.jetbrains.skiko.LibraryTestImpl.run(Library.kt:101)
```

Reply:
The last exception is fixed by:
```
val ext = if (hostOs.isWindows) ".dll" else ""
val tempFile = File.createTempFile("skiko", ext)
```
Test doesn't fail after that

Reply:
Thanks, hopefully covered all the issues. The beauty of testing :).